File: perl/WebView.pl
use Plack::Builder;
use Plack::Request;
use Template;
use Redis;

sub arr_in_arr {
	my $in_arr = shift;

	my @arr_in_arr;
	my @arr;
	my $c = 0;
	for my $a (@{$in_arr}) {
		if(($c+=1)%2){
			@arr = ();
			push(@arr,$a);
		}else{
			push(@arr,$a);
			push(@arr_in_arr,[@arr]);
		}
	}

	return @arr_in_arr;
};

my $app = sub {
	my $env = shift;

	my $redis = Redis->new(server => "127.0.0.1:6379");

	my @total_num = $redis->zrevrange('total_num',0,-1,'withscores');
	my @num = $redis->zrevrange('num',0,-1,'withscores');

	@total_num = arr_in_arr(\@total_num),
	@num = arr_in_arr(\@num),

	my $pages = 0;
	#foreach my @key (@total_num) {
	#	$pages += $key[1];
	#}
	foreach my $key (@total_num) {
		$pages += ${$key}[1];
	}

	my $start_time = $redis->get('start_time');
	my $finish_time = $redis->get('finish_time');
	my $duration = $finish_time - $start_time;

	my $vars =  +{
		pages => $pages,
		duration => $duration,
		total_num => \@total_num,
		num => \@num,
	};

	my $tt = new Template;
	$tt->process('views/index.html', $vars , \my $html) or die $tt->error;

	return [ 200, [ "Content-Type" => "text/html" ], [ $html ] ];

};

builder {
	mount "/" => $app;
};

File: node/WebView.js
var express = require('express');
var client = require('redis').createClient();
var app = express();
var ECT = require('ect');
app.engine('ect', ECT({ watch: true, root: __dirname+'/views', ext: '.ect'}).render);
app.set('view engine','ect');

function is(type, obj) {
  var clas = Object.prototype.toString.call(obj).slice(8, -1);
  return obj !== undefined && obj !== null && clas === type;
}

function map_to_int(arr){
  var resarr = []
  arr.forEach(function(elem){
    resarr.push(parseInt(elem))
  })
  return resarr
}

function each_slice(arr,gather){
  if(is("Number",gather)){
    var cnt=0;
    var resarr = [];
    var elemarr = [];
    arr.forEach(function(elem){
      elemarr.push(elem)
      if(cnt%gather==1){
        resarr.push(elemarr);
        elemarr = [];
      }
      cnt += 1;
    })
  }
  return resarr;
}

function sum_with_gap(arr,start,gap){
  var cnt=0;
  var sum=0;
  arr.forEach(function(elem){
    if(cnt>=start && cnt%gap==start){
      sum += parseInt(elem)
    }
    cnt += 1
  })
  return sum
}

app.get('/', function (req, res) {
  client.zrevrange("total_num", 0, -1, 'withscores', function(err, total_num){
    client.zrevrange("num", 0, -1, 'withscores', function(err, num){
      client.get("start_time", function(err, start_time){
        client.get("stop_time", function(err, stop_time){
          pages = sum_with_gap(total_num,1,2)
          res.render('index', {
            pages: pages,
            duration: (stop_time-start_time)/1000,
            throughput: pages/((stop_time-start_time)/1000)+"",
            total_num: each_slice(map_to_int(total_num),2),
            num: each_slice(map_to_int(num),2),
          });
        });
      });
    });
  });
});

app.listen(3000);

